k8s/persistent_volumes.go

```diff
@@ -51,3 +62,161 @@ func DeletePersistentVolumeClaims(ctx context.Context, namespaceOverride string,
 
 	return client.CoreV1().PersistentVolumeClaims(namespace).DeleteCollection(ctx, metav1.DeleteOptions{}, listOptions)
 }
+
+var TarImage = "quay.io/boson/alpine-socat:1.7.4.3-r1-non-root"
+
+// UploadToVolume uploads files (passed in form of tar stream) into volume.
+func UploadToVolume(ctx context.Context, content io.Reader, claimName, namespace string) error {
+	return runWithVolumeMounted(ctx, TarImage, []string{"tar", "-xmf", "-"}, content, claimName, namespace)
+}
+
+// Runs a pod with given image, command and stdin
+// while having the volume mounted and working directory set to it.
+func runWithVolumeMounted(ctx context.Context, podImage string, podCommand []string, podInput io.Reader, claimName, namespace string) error {
+	var err error
+
+	cliConf := GetClientConfig()
+	restConf, err := cliConf.ClientConfig()
+	if err != nil {
+		return fmt.Errorf("cannot get client config: %w", err)
+	}
+	restConf.WarningHandler = restclient.NoWarnings{}
+
+	err = setConfigDefaults(restConf)
+	if err != nil {
+		return fmt.Errorf("cannot set config defaults: %w", err)
+	}
+
+	client, err := kubernetes.NewForConfig(restConf)
+	if err != nil {
+		return fmt.Errorf("cannot create k8s client: %w", err)
+	}
+
+	namespace, err = GetNamespace(namespace)
+	if err != nil {
+		return fmt.Errorf("cannot get namespace: %w", err)
+	}
+
+	podName := "volume-uploader-" + rand.String(5)
+
+	pods := client.CoreV1().Pods(namespace)
+
+	defer func() {
+		_ = pods.Delete(ctx, podName, metav1.DeleteOptions{})
+	}()
+
+	const volumeMntPoint = "/tmp/volume_mnt"
+	const pVol = "p-vol"
+	pod := &corev1.Pod{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:        podName,
+			Labels:      nil,
+			Annotations: nil,
+		},
+		Spec: corev1.PodSpec{
+			Containers: []corev1.Container{
+				{
+					Name:       podName,
+					Image:      podImage,
+					Stdin:      true,
+					StdinOnce:  true,
+					WorkingDir: volumeMntPoint,
+					Command:    podCommand,
+					VolumeMounts: []corev1.VolumeMount{
+						{
+							Name:      pVol,
+							MountPath: volumeMntPoint,
+						},
+					},
+					SecurityContext: defaultSecurityContext(),
+				},
+			},
+			Volumes: []corev1.Volume{{
+				Name: pVol,
+				VolumeSource: corev1.VolumeSource{
+					PersistentVolumeClaim: &corev1.PersistentVolumeClaimVolumeSource{
+						ClaimName: claimName,
+					},
+				},
+			}},
+			DNSPolicy:     corev1.DNSClusterFirst,
+			RestartPolicy: corev1.RestartPolicyNever,
+		},
+	}
+
+	localCtx, cancel := context.WithCancel(ctx)
+	defer cancel()
+	ready := podReady(localCtx, client.CoreV1(), podName, namespace)
+
+	_, err = pods.Create(ctx, pod, metav1.CreateOptions{})
+	if err != nil {
+		return fmt.Errorf("cannot create pod: %w", err)
+	}
+
+	select {
+	case err = <-ready:
+	case <-ctx.Done():
+		err = ctx.Err()
+	case <-time.After(time.Minute * 5):
+		err = errors.New("timeout")
```

Comment:
Nit: it might be nice for the user to have slightly more detail here
```suggestion
		err = errors.New("timeout waiting for pod to start")
```